### PR TITLE
WIP: Fix to tick scoring function

### PR DESCRIPTION
**Pull request by crbinz**
Fixes my bug where I was getting only a single tick on an axis for a specific data set, while (hopefully) making minimal changes. Comments welcome.

The test case resulting in a single tick is

```
scatter(s[:,1], s[:,2]) # ticks are fine
plot!(e1[:,1], e1[:,2]) # single y axis tick
```
where `s` and `e1` are defined [here](https://gist.github.com/crbinz/efdcfe4b99aa6430ebf021cf9240bde3)

### CODE REVIEW

**Comment by crbinz** [comment]
The test case resulting in a single tick is

```
scatter(s[:,1], s[:,2]) # ticks are fine
plot!(e1[:,1], e1[:,2]) # single y axis tick
```
where `s` and `e1` are defined [here](https://gist.github.com/crbinz/efdcfe4b99aa6430ebf021cf9240bde3)

**Comment by daschw** [comment]
Hi @crbinz, nice work and thanks for putting so much effort into this!

Admittedly, I did not find an example, where your changes caused different results other than your degenerate case, you posted in the Gitter channel. However I have a proposal, how less of the functions logic could change:

You changed
```
if strict_span && span > xspan
    score -= 10000
elseif !strict_span && (span >= 2.0*xspan || span < xspan)
    score -= 1000
end
```
to
```
if span >= 2.0*xspan
    score -= 1000
end
```
Since this function is called from Plots with `strict_span = true`, this means now every case (instead of none) is checked for `span >= 2.0*xspan` and `span == xspan` is not rewarded any more.

In your degenerate example axis limits were approx. 2.75 and 3.75, so `xspan` was roughly 1, and the ticks chosen by the algorithm before applying `strict_span` (https://github.com/JuliaPlots/PlotUtils.jl/blob/master/src/ticks.jl#L163-L173) were `[0.0:5.0]`, so span was 5.

Thus, keeping
```
if strict_span && span > xspan
    score -= 10000
elseif !strict_span && (span >= 2.0*xspan || span < xspan)
    score -= 1000
end
```
as it is and adding, e.g.
```
if span >= 4.0*xspan
    score -= 1000
end
```
below, would also resolve your case and probably affect fewer other cases.

The question is, if we want to allow cases, where `span > 2*xspan ` or `span > 3*xspan `?

Anyway, as I said, I did not find an example, where your modification affected any other plot, and I am by no means a ticks algorithm expert, so let me know, what you think!

File: src/ticks.jl [hunk 1]
```diff
@@ -114,17 +114,15 @@ function optimize_ticks_typed{T}(x_min::T, x_max::T, extend_ticks,
                     g = 0 < k < 2k_ideal ? 1 - abs(k - k_ideal) / k_ideal : 0.0
 
                     # coverage
-                    c = 1.5 * xspan/span
+                    c = 1.5*xspan/span
```
Comment (mkborregaard) [inline src/ticks.jl]: Also please revert this unrelated change :-)

**Comment by daschw** [comment]
For anyone watching this:
before:
onetick

afterwards:
moreticks

File: src/ticks.jl (hunk 1 above)
Comment (crbinz) [inline src/ticks.jl]: Will do

**Comment by crbinz** [comment]
@daschw thank you for the comments! I want to think about your proposal, but at first glance it seems better in that it will potentially change the existing behavior less. The differences I mentioned were in running the VisualRegressionTests, fyi.

My problem with the first condition (`if strict_span && span > xspan`) is that  because of [this check](https://github.com/JuliaPlots/PlotUtils.jl/blob/master/src/ticks.jl#L97), it will always evaluate to `true` (except in the case where `span == xspan`, which seems quite unlikely to me). 

Also, considering `elseif !strict_span && (span >= 2.0*xspan || span < xspan)` will never be `true` (at least, the way it's currently called from `Plots`), would it be better to remove?

**Comment by daschw** [comment]
> The differences I mentioned were in running the VisualRegressionTests, fyi

OK, thanks, to be honest, I did not do that much checking. Maybe you could check whether my proposal results in differences (w.r.t. tick positions) in the VRT tests?
> My problem with the first condition (if strict_span && span > xspan) is that because of this check, it will always evaluate to true (except in the case where span == xspan, which seems quite unlikely to me).

> Also, considering elseif !strict_span && (span >= 2.0*xspan || span < xspan) will never be true (at least, the way it's currently called from Plots)

I totally agree with you on these points: Considering the calls from Plots' `optimal_ticks_and_labels` with the default values this condition really almost never has an effect - it would just reward the very unlikely cases of `span == xspan`.
However, I don't see, why we should remove it then. From the comments in the file this seems like a generic implementation of a probably well-known scoring algorithm and the contributing section in the Plots docs says about PlotUtils:
> Components that could be used for other (non-Plots) packages. Anything that is sufficiently generic and useful could be contributed here.

Maybe someone wants to use this algorithm with other defaults for another purpose ...
Hence, I would rather change as little as possible.

**Comment by crbinz** [comment]
> Maybe you could check whether my proposal results in differences (w.r.t. tick positions) in the VRT tests?

I will definitely run the tests on whatever we land on here 😄 

> From the comments in the file this seems like a generic implementation of a probably well-known scoring algorithm

From what I could tell, this implementation is a modified version of a scoring algorithm originally devised by Wilkinson. The original explanation only includes the scores for niceness, simplicity, granularity, and coverage; and none of these extra checks. That said, presumably these checks were added for some good reason, and I don't want to discard any hard-fought knowledge they represent. I do wish someone could explain it to me, though... (@dcjones, perhaps?)

This is all to say: I see your point, and I agree. As to what the factor should be (2, 3, 4?), the [reference](http://vis.stanford.edu/files/2010-TickLabels-InfoVis.pdf) I read recommends no more than 20% whitespace, which would dictate a factor of `5` (right?).

**Comment by daschw** [comment]
>  no more than 20% whitespace, which would dictate a factor of 5 (right?)

I don't think , you can really say that this way. Here's, how I see the bahavior for Plots: We have `strict_span == true` (which makes sense, because we don't want to change possible user-set axis limits with the ticks algorithm). Thus, in the end only the ticks between `x_min` and `x_max` are returned. The factor (2, 3, 4, ...) would define the maximal axes span, which the `kmin = 5` to `kmax = 8` initially determined ticks may be spread over. You found a case, where the chosen ticks (only 5) were spread over a span about 5 times larger than `x_max - x_min`. And in this case only one tick was returned finally.

I would summarize it like this (correct, me if I'm wrong):
- If a smaller factor is chosen, more ticks will be returned in general and the probability to run into another degenerate example is lower.
- If a higher factor is chosen, fewer of already created plots are affected by this modification, because before span was not limited at all.

I would go for a maximal factor, that still resolves your case, which will probably (hopefully) barely affect any other cases.

File: src/ticks.jl [hunk 2]
```diff
@@ -114,17 +114,15 @@ function optimize_ticks_typed{T}(x_min::T, x_max::T, extend_ticks,
...
 
                     score = granularity_weight * g +
                             simplicity_weight * s +
                             coverage_weight * c +
                             niceness_weight * qscore
 
-                    # strict limits on coverage
-                    if strict_span && span > xspan
-                        score -= 10000
```
Comment (daschw) [inline src/ticks.jl]: Adding
```
if span >= α*xspan
    score -= 1000
end
```
here should be sufficient for your case (depending on α ... 4 works, 4.5 is already too big) and would not affect the `strict_span == false` case

**Comment by crbinz** [comment]
> If a higher factor is chosen, fewer of already created plots are affected by this modification, because before span was not limited at all.

Well put. It's a bit of a band-aid, but considering this issue hasn't been raised before my test case (that I could tell), maybe that's all that is needed. My vote, in that case, would be for `α = 4.0`.

**Comment by crbinz** [comment]
Reimplemented this change as discussed. Running the VisualRegressionTests, I noticed no changes in tick spacing.
